assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,432 @@
+        self.spans.push(vec_with_iter_capacity(iter));
+    }
+
+    fn append_to_span(&mut self, operation: Operation) {
+        if let Some(span) = self.spans.last_mut() {
+            span.push(operation);
+        }
+    }
+
+    fn end_span(&mut self) {
+        self.spans
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(CodeBlock::new_span)
+            .map(|span| self.blocks.last_mut().map(|block| block.push(span)));
+    }
+
+    fn start_block<I, T>(&mut self, iter: &I)
+    where
+        I: Iterator<Item = T>,
+    {
+        self.blocks.push(vec_with_iter_capacity(iter));
+    }
+
+    fn pop_block(&mut self) -> Option<CodeBlock> {
+        self.blocks
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(parsers::combine_blocks)
+    }
+
+    fn append_block(&mut self, block: CodeBlock) {
+        self.blocks.last_mut().map(|blocks| blocks.push(block));
+    }
+
+    fn compile_body<A, N>(&mut self, ast: A) -> CodeBlock
+    where
+        A: Iterator<Item = N>,
+        N: Borrow<Node>,
+    {
+        self.start_block(&ast);
+        self.start_span(&ast);
+
+        for node in ast {
+            match node.borrow() {
+                Node::Instruction(instruction) => {
+                    self.compile_instruction(instruction);
+                }
+
+                Node::IfElse(t, f) => {
+                    let t = self.compile_body(t.iter());
+                    let f = self.compile_body(f.iter());
+
+                    let block = CodeBlock::new_split(t, f);
+
+                    self.append_block(block);
+                }
+
+                Node::Repeat(n, nodes) => {
+                    let block = self.compile_body(nodes.iter());
+
+                    for _ in 0..*n {
+                        self.append_block(block.clone());
+                    }
+                }
+
+                Node::While(nodes) => {
+                    let block = self.compile_body(nodes.iter());
+                    let block = CodeBlock::new_loop(block);
+
+                    self.append_block(block);
+                }
+            }
+        }
+
+        self.end_span();
+        self.pop_block().unwrap()
+    }
+
+    fn compile_instruction(&mut self, instruction: &Instruction) {
+        use Operation::*;
+
+        match instruction {
+            Instruction::Assert => self.append_to_span(Assert),
+            Instruction::AssertEq => {
+                self.append_to_span(Eq);
+                self.append_to_span(Assert);
+            }
+            Instruction::Assertz => {
+                self.append_to_span(Eqz);
+                self.append_to_span(Assert);
+            }
+            Instruction::Add => self.append_to_span(Add),
+            Instruction::AddImm(imm) if imm == &Felt::ONE => self.append_to_span(Incr),
+            Instruction::AddImm(imm) => {
+                self.append_to_span(Push(*imm));
+                self.append_to_span(Add);
+            }
```

Comment:
For things like this I'd prefer to create a separate helper function (otherwise this will become very difficult to read). Something like:
```Rust
fn append_add_imm(imm: Felt, span_ops: &mut Vec<Operation>) {
    if imm == ONE {
        span_ops.push(Incr);
    }
    else {
        span_ops.push(Push(imm));
        span_ops.push(Add);
    }
}
```

Reply:
Refactored into modules